File: cov.py
#!/usr/bin/env python3

"""
    File to test the conversion from quaternion covariances to euler angles
"""
import numpy as np

def convert(quat, quatcov):
    """
        Function to do the conversion
    """
    jac = jacobian(quat)
    eul_cov = jac @ quatcov @ jac.T
    return eul_cov

def jacobian(quat):
    """
        Function to return the jacobian - as defined by:
        https://stats.stackexchange.com/questions/119780/what-does-the-covariance-of-a-quaternion-mean
    """
    jac = np.zeros((3,4))
    q32n = quat[2] - quat[1]
    q32p = quat[2] + quat[1]
    q41p = quat[3] + quat[0]
    q41n = quat[3] - quat[0]
    normalizerpos = q32p**2 + q41p**2
    normalizerneg = q32n**2 + q41n**2

    q41pos = q41p / normalizerpos
    q32pos = q32p / normalizerpos
    q32neg = q32n / normalizerneg
    q41neg = q41n / normalizerneg
    
    # Row 1 of the jacobian
    psi1 = (q32pos*-1.0) + (q32neg)
    psi2 = (q41pos) - (q41neg)
    psi3 = (q41pos) + (q41neg)
    psi4 = (q32pos*-1.0) - (q32neg)

    # Row 3 of the Jacobian
    phi1 = psi4
    phi2 = psi3
    phi3 = psi2
    phi4 = psi1

    # Row 2 of the Jacobian
    theta_norm = 2 / np.sqrt(1 - 4 * ((quat[1] * quat[2] + quat[0] * quat[3])**2)) 

    theta1 = theta_norm * quat[3]
    theta2 = theta_norm * quat[2]
    theta3 = theta_norm * quat[1]
    theta4 = theta_norm * quat[0]

    # filling in the jacobian
    jac[0,0] = psi1
    jac[0,1] = psi2
    jac[0,2] = psi3
    jac[0,3] = psi4

    jac[1,0] = theta1
    jac[1,1] = theta2
    jac[1,2] = theta3
    jac[1,3] = theta4

    jac[2,0] = phi1
    jac[2,1] = phi2
    jac[2,2] = phi3
    jac[2,3] = phi4

    return jac

def randomQuat():
    """
        Following The definition used by Eigen
        Linked here: http://planning.cs.uiuc.edu/node198.html
    """
    unitquat = np.zeros(4)
    u = np.random.rand(3)
    unitquat[0] = np.sqrt(1-u[0]) * np.sin(2*np.pi*u[1])
    unitquat[1] = np.sqrt(1-u[0]) * np.cos(2*np.pi*u[1])
    unitquat[2] = np.sqrt(u[0]) * np.sin(2*np.pi*u[2])
    unitquat[3] = np.sqrt(u[0]) * np.cos(2*np.pi*u[2])
    unitquat = unitquat / np.linalg.norm(unitquat)
    return unitquat


if __name__=="__main__":
    # seed the random generator
    np.random.seed(1)

    # dummy covariance
    quatcov = np.ones((4,4))

    quatlist = []
    for i in range(1000):
        quat = randomQuat()
        eucov = convert(quat, quatcov)
    print("completed all random test cases. Proceeding")
    
    quat = np.ones(4) / np.linalg.norm(np.ones(4))
    print("Attempting numerically instable version: {}".format(quat))

    eucov = convert(quat, quatcov)
    print(eucov)

    

    

File: cov.cpp
// #include <eigen3/Eigen/Dense>
#include <Eigen/Dense>
#include <iostream>

using namespace Eigen;

void jacobian(MatrixXd& , Quaterniond);

void convert(Matrix3d& eul_cov, Quaterniond quat, const Matrix4d quatcov){

    MatrixXd jac = MatrixXd::Zero(3,4);
    jacobian(jac, quat);

    eul_cov = jac * quatcov * jac.transpose();
}

void jacobian(MatrixXd& jac, Quaterniond quat){
    // convention in equation 2.8, p. 12 of the thesis
    double q32n = quat.z() - quat.y();
    double q32p = quat.z() + quat.y();
    double q41p = quat.w() + quat.x();
    double q41n = quat.w() - quat.x();

    double normalizerp = q32p*q32p + q41p*q41p;
    double normalizern = q32n*q32n + q41n*q41n;

    double q41pos = q41p / normalizerp;
    double q32pos = q32p / normalizerp;
    double q41neg = q41n / normalizern;
    double q32neg = q32n / normalizern;

    double theta_norm = 2 / sqrt(1.f - 4.f * (quat.x() * quat.y() + quat.z() * quat.w()) * (quat.x() * quat.y() + quat.z() * quat.w())); 
    // Matrix values - simplified
    double psi1 = q32pos * -1.f + q32neg;
    double psi2 = q41pos - q41neg;
    double psi3 = q41pos + q41neg;
    double psi4 = q32pos * -1.f - q32neg;

    // filling in the matrix - row 1
    jac(0,0) = psi1;
    jac(0,1) = psi2;
    jac(0,2) = psi3;
    jac(0,3) = psi4;

    // Row 3
    jac(2,0) = psi4;
    jac(2,1) = psi3;
    jac(2,2) = psi2;
    jac(2,3) = psi1;

    // Row 2
    jac(1,0) = theta_norm * quat.z();
    jac(1,1) = theta_norm * quat.y();
    jac(1,2) = theta_norm * quat.x();
    jac(1,3) = theta_norm * quat.w();
        
}

int main(int argc, char** argv){

    Matrix4d quatcov = MatrixXd::Ones(4,4);
    Quaterniond quat;
    Matrix3d eul_cov;

    for (int i=0; i < 100; ++i){
        quat = Quaterniond::UnitRandom(); //initialization order is important, see doc. w, x, y, z. Internal storage is xyz, w
        convert(eul_cov, quat, quatcov);
        // std::cout << "Conversion for \nw: " << quat.w() << " x: " << quat.x() << " y: " <<quat.y() << " z: " << quat.z() << "\nWorks" << std::endl;
        // std::cout<<eul_cov << std::endl;
    }

    // Edge case:
    quat = Quaterniond(0.5, 0.5, 0.5, 0.5);
    convert(eul_cov, quat, quatcov);    
    // std::cout<<eul_cov<<std::endl;   

    return 0;
}